src/utils.ts

```diff
@@ -303,3 +311,27 @@ export async function createUnsignedFillRelayTransaction(
     { from: simulatedRelayerAddress }
   );
 }
+
+/**
+ * Calculates the average block production speed over the past ~1 day.
+ * @param provider A valid ethers provider - will be used to calculate block production speed.
+ */
+export async function calculateBlockSpeed(
+  provider: providers.Provider | L2Provider<providers.Provider>
+): Promise<number> {
+  const latestBlock = await provider.getBlock("latest");
+  const latestBlockTimestamp = latestBlock.timestamp;
+  const chainId = (await provider.getNetwork()).chainId;
+  const sampleSpeed = BlockSpeedSample[chainId];
+  let lastBlockNumber = latestBlock.number;
+  let lastBlockTimestamp = latestBlockTimestamp;
+  // Keep searching for the first block older than 1 day.
+  // This is only a rough estimate and can go back further than 1 day. However, this would minimize the number
+  // of requests and is generally accurate unless the rate of block production has been very volatile over a
+  // short period of time.
+  while (latestBlockTimestamp - lastBlockTimestamp < ONE_DAY_SECS) {
```

Comment:
It might not be long enough. This is just a simple heuristic-based approach - if it's long enough, keep searching back. Ideally this can be more sophisticated by conducting a search similar to binary search to find the closest block to ~1 day from the latest. However, this might not reduce the number of requests as much and is just a lot more complex

Reply:
ah I see I didn't fully understand the algo at first